inform7/Internal/Inter/Architecture32Kit/Sections/Glk.i6t

```diff
@@ -4,1001 +4,499 @@ Glk is a fully-featured input-output model, designed by Andrew Plotkin for use
+		len = win.line_input_buffer_maxlen;
+	}
+	for (i = 0: i < len: i++) {
+		ch = BlkValueRead(txt, i);
+		if (ch == 0) break;
+		if (uni) {
+			buf-->i = ch;
+		}
+		else if (ch <= $FF) {
+			buf->i = ch;
+		}
+		else {
+			buf->i = '?';
+		}
+	}
+	TEXT_TY_Untransmute(txt, p, cp);
+	return i;
 ];
 
-[ glk_put_char_stream_uni _vararg_count;
-  ! glk_put_char_stream_uni(stream, uint)
-  @glk 299 _vararg_count 0;
-  return 0;
+[ FindGlkWindowFromRefNum ref win;
+	if (ref ~= 0) {
+		for (win = K_Glk_Window_First: win: win = win.K_Glk_Window_Next) {
+			if (win.glk_ref == ref) {
+				return win;
+			}
+		}
+	}
+	return nothing;
 ];
 
-[ glk_put_string_stream_uni _vararg_count;
-  ! glk_put_string_stream_uni(stream, unicode)
-  @glk 300 _vararg_count 0;
-  return 0;
+[ FillInWindowDetails win ref;
+	win.glk_ref = ref;
+	win.glk_rock = glk_window_get_rock(ref);
+	win.glk_window_type = glk_window_get_type(ref);
 ];
 
-[ glk_put_buffer_stream_uni _vararg_count;
-  ! glk_put_buffer_stream_uni(stream, uintarray, arraylen)
-  @glk 301 _vararg_count 0;
-  return 0;
+[ GetWindowSize win index;
+	glk_window_get_size(win.glk_ref, gg_arguments, gg_arguments + WORDSIZE);
+	return gg_arguments-->index;
 ];
 
-[ glk_get_char_stream_uni _vararg_count ret;
-  ! glk_get_char_stream_uni(stream) => int
-  @glk 304 _vararg_count ret;
-  return ret;
-];
+@h Glk events.
+To handle events we intercept the |glk_select| function. This allows us to handle
+events early and consistently.
 
-[ glk_get_buffer_stream_uni _vararg_count ret;
-  ! glk_get_buffer_stream_uni(stream, uintarray, arraylen) => uint
-  @glk 305 _vararg_count ret;
-  return ret;
-];
-
-[ glk_get_line_stream_uni _vararg_count ret;
-  ! glk_get_line_stream_uni(stream, uintarray, arraylen) => uint
-  @glk 306 _vararg_count ret;
-  return ret;
-];
+=
+Global Glk_Event_Struct_type;
+Global Glk_Event_Struct_win;
+Global Glk_Event_Struct_val1;
+Global Glk_Event_Struct_val2;
+
+[ glk_select event_struct;
+	! Call the real glk_select
+	@push event_struct;
+	@glk 192 1 0;
+
+	! Clear the unknown window's ref so that we don't immediately match it
+	Unknown_Glk_Window.glk_ref = 0;
+
+	! Copy values to our variables
+	Glk_Event_Struct_type = event_struct-->0;
+	Glk_Event_Struct_win = event_struct-->1;
+	Glk_Event_Struct_val1 = event_struct-->2;
+	Glk_Event_Struct_val2 = event_struct-->3;
+
+	! Turn a window reference ID into a window object
+	if (Glk_Event_Struct_win > 0) {
+		Glk_Event_Struct_win = FindGlkWindowFromRefNum(Glk_Event_Struct_win);
+		! If no window was found, then use the unknown window
+		if (Glk_Event_Struct_win == nothing) {
+			Glk_Event_Struct_win = Unknown_Glk_Window;
+			FillInWindowDetails(Unknown_Glk_Window, event_struct-->1);
+		}
 
-[ glk_stream_open_file_uni _vararg_count ret;
-  ! glk_stream_open_file_uni(fileref, uint, uint) => stream
-  @glk 312 _vararg_count ret;
-  return ret;
-];
+		! If the current event type is line input, copy the buffer so I7 can access it
```

Comment:
I'm rethinking this now, it is very good in terms of the experience for authors, but the performance hit each and every turn is not insignificant. I think the text conversion should be manual instead.